### ISSUE TITLE
3.3-Pytorch-Dataset& Dataloader

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#8]
### Working with data
![image](https://user-images.githubusercontent.com/34474924/195294223-4bb3b9c9-e527-4334-ab98-32ae7e5856f2.png)
要將data先load進去才可以進行training等步驟
`Dataset`: stores data samples and expected values
`Dataloader`: groups data in batches, enables multiprocessing

PyTorch domain libraries provide a number of pre-loaded datasets (such as FashionMNIST)

Here is an example of how to load the Fashion-MNIST dataset from TorchVision
```python
import torch
from torch.utils.data import Dataset
from torchvision import datasets
from torchvision.transforms import ToTensor
import matplotlib.pyplot as plt


training_data = datasets.FashionMNIST(
    root="data",
    train=True,
    download=True,
    transform=ToTensor()
)

test_data = datasets.FashionMNIST(
    root="data",
    train=False,
    download=True,
    transform=ToTensor()
)
```

`root` is the path where the train/test data is stored,

`train` specifies training or test dataset,

`download=True` downloads the data from the internet if it’s not available at root.

`transform` and `target_transform` specify the feature and label transformations

### Iterating and Visualizing the Datase
We can index Datasets manually like a list: `training_data[index]`. We use `matplotlib` to visualize some samples in our training data.

_____________________________
### Creating a Custom Dataset for your files
A custom Dataset class must implement three functions: `__init__`, `__len__`, and `__getitem__`. 
the FashionMNIST images are stored in a directory img_dir, and their labels are stored separately in a CSV file annotations_file.

In the next sections, we’ll break down what’s happening in each of these functions.

```python
dataset = MyDataset(file)
dataloader = DataLoader(dataset, batch_size, shuffle=True)
//Training: shuffle=True
//Training: shuffle=False
```
